staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go

```diff
@@ -221,27 +224,38 @@ func (cnc *CloudNodeController) reconcileNodeLabels(nodeName string) error {
 }
 
 // UpdateNodeAddress updates the nodeAddress of a single node
-func (cnc *CloudNodeController) updateNodeAddress(ctx context.Context, node *v1.Node, instances cloudprovider.Instances) {
+func (cnc *CloudNodeController) updateNodeAddress(ctx context.Context, node *v1.Node, instances cloudprovider.Instances, instancesV2 cloudprovider.InstancesV2) {
 	// Do not process nodes that are still tainted
 	cloudTaint := getCloudTaint(node.Spec.Taints)
 	if cloudTaint != nil {
 		klog.V(5).Infof("This node %s is still tainted. Will not process.", node.Name)
 		return
 	}
-	// Node that isn't present according to the cloud provider shouldn't have its address updated
-	exists, err := ensureNodeExistsByProviderID(ctx, instances, node)
-	if err != nil {
-		// Continue to update node address when not sure the node is not exists
-		klog.Errorf("%v", err)
-	} else if !exists {
-		klog.V(4).Infof("The node %s is no longer present according to the cloud provider, do not process.", node.Name)
-		return
-	}
 
-	nodeAddresses, err := getNodeAddressesByProviderIDOrName(ctx, instances, node.Spec.ProviderID, node.Name)
-	if err != nil {
-		klog.Errorf("Error getting node addresses for node %q: %v", node.Name, err)
-		return
+	var nodeAddresses []v1.NodeAddress
+	if instancesV2 != nil {
```

Comment:
I think we can make this and the code in getNodeModifiersFromCloudProvider a bit cleaner by having one method that returns InstanceMetadata. If `instancesV2` is implemented, construct `InstanceMetadata` from `InstanceMetadataByProviderID`, if `instances` is implemented, construct `InstanceMetadata` by calling the individual methods in `Instances`.

Reply:
Adopted your suggestion and implemented a method getInstanceMetadata which used by getNodeModifiersFromCloudProvider. While getInstanceTypeByProviderIDOrName isn't called here. A redundant call will be produced if replace these code with getInstanceMetadata.

Reply:
Can we have a separate method for the addresses then? Want to hide the instance / instance v2 checks from the main code path and keep that hidden in separate methods.

Reply:
I also want to hide instances/instancesV2 checks in `initializeNode`, however updateNodeAddress and getNodeModifiersFromCloudProvider's logic are slightly different. Do you mean `updateNodeAddress` and `getNodeModifiersFromCloudProvider` using different method to get instanceMetadata(in instances() case)?

Reply:
>  Do you mean updateNodeAddress and getNodeModifiersFromCloudProvider using different method to get instanceMetadata(in instances() case)?

yes, at least for now.

Reply:
Updated, PTAL.